pkg/cache/resource_node.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"maps"
+
+	"sigs.k8s.io/kueue/pkg/resources"
+)
+
+type ResourceNode struct {
```

Comment:
I think you might be able to add one of the hierarchy objects as a member of this object, so that the functions below are members of this struct, but first it might be worth applying some of the remarks I left in #2824.